youtrack-backup.pl

```diff
@@ -0,0 +1,195 @@
+$result = `$curl_cmd | jq '.backupStatus.backupError'`;
+
+chomp($result);
+if ($result ne "null") {
+	die "Database backup failed: $result\n";
+}
+
+#
+# Wait until the database backup completes
+#
+$jq_cmd     = "jq '.backupInProgress'";
+$curl_query = "--request GET '${yt_url}/api/admin/databaseBackup/settings/backupStatus?fields=backupInProgress'";
+$curl_cmd   = "curl $accept_header $auth_header $progress_header $curl_query | $jq_cmd";
+
+while (1) {
+	sleep($delay);
+	$result = `$curl_cmd`;
+	chomp $result;
+	if ($result eq "true") {
+		printf "%s: waiting for DB backup to complete...\n", tstamp() if (! $quiet);
+	}
+	else {
+		last;
+	}
+}
+
+#
+# Get a link to the database backup file
+#
+$jq_cmd     = "jq '.[].id,.[].link'";
+$curl_query = "--request GET '${yt_url}/api/admin/databaseBackup/backups?fields=creationDate,link,id,size'";
+$curl_cmd   = "curl $accept_header $auth_header $progress_header $curl_query | $jq_cmd";
+
+printf "%s: Getting download link\n", tstamp() if (! $quiet);
+$result = `$curl_cmd`;
+chomp $result;
+$result =~ s/"//g;	# remove double quotes
+
+my $file_name;
+my $download_path;
+my $download_url;
+($file_name,$download_url) = split(/\n/, $result);
+$download_path = "/tmp/$file_name";
+
+
+#
+# Remove '\r\n' from the link's signature.  The characters don't belong
+# and are believed to be a bug in JetBrain's server-side code.  There
+# are two instances of the four-character string '\r\n'.
+#
+$download_url =~ s/\\r\\n//g;
+
+
+#
+# Compose the complete download URL
+#
+$download_url = "${yt_url}/" . $download_url;
+
+#
+# Download the database backup file
+#
+$curl_query = "--request GET $download_url --output $download_path";
+$curl_cmd   = "curl $accept_header $auth_header $progress_header $curl_query";
+
+printf "%s: Downloading file $download_path\n", tstamp() if (! $quiet);
+$result = system($curl_cmd);
+if ($result != 0) {
+	die "Download failed: $result\n";
+}
+
+#
+# Copy the database backup file to the Backblaze B2 bucket
+#
+# References:
+# https://www.backblaze.com/apidocs/b2-upload-file
+# https://www.backblaze.com/docs/cloud-storage-quick-start-guides
+# https://www.backblaze.com/docs/cloud-storage-developer-quick-start-guide
+#
+# CLI tools
+# https://www.backblaze.com/docs/cloud-storage-command-line-tools
+# https://www.backblaze.com/docs/cloud-storage-upload-files-with-the-cli
+# https://f000.backblazeb2.com/file/jsonwaterfalls/B2%20CLI%20Guide.pdf
+
+printf "%s: Uploading file to Backblaze B2 bucket $b2_bucket\n", tstamp() if (! $quiet);
+my $tmpfile = File::Temp->new( TEMPLATE => '/tmp/XXXXXXXXXX' );
+my $cmd     = "b2 upload-file --noProgress $b2_bucket $download_path $file_name > $tmpfile";
+$result     = system($cmd);
+if ($result != 0) {
+	`cat $tmpfile`;
+	unlink($download_path);
+	die "Upload failed: $result\n";
+}
+
+printf "%s: Upload to Backblaze B2 bucket complete\n", tstamp() if (! $quiet);
+
+unlink($download_path);
+
+exit(0);
```

Comment:
As you probably noticed, I don't have much to say about the Perl code. I'm trusting you've tested it. Speaking of testing, is it possible to write any unit tests for this Perl code?

Reply:
The Perl code has worked several times.  Don't know how I would perform unit tests on it.